src/toil/lib/misc.py

```diff
@@ -191,3 +192,62 @@ def truncExpBackoff():
         t *= 2
     while True:
         yield t
+
+
+def atomic_tmp_file(final_path):
+    """Return a tmp file name to use with atomic_install.  This will be in the
+    same directory as final_path. The temporary file will have the same extension
+    as finalPath.  In final path is in /dev (/dev/null, /dev/stdout), it is
+    returned unchanged and atomic_tmp_install will do nothing."""
+    final_dir = os.path.dirname(os.path.normpath(final_path))
+    if final_dir == '/dev':
+        return final_path
+    if final_dir == "":
+        final_dir = '.'
+    final_basename = os.path.basename(final_path)
+    final_ext = os.path.splitext(final_path)[1]
+    base_name = "{}.{}.{}.tmp{}".format(final_basename, socket.gethostname(), os.getpid(), final_ext)
+    return  os.path.join(final_dir, base_name)
+
+
+def atomic_install(tmp_path, final_path):
+    "atomic install of tmp_path as final_path"
+    if os.path.dirname(os.path.normpath(final_path)) != '/dev':
+        os.rename(tmp_path, final_path)
+
+
+class AtomicFileCreate(object):
+    """Context manager to create a temporary file.  Entering returns path to
+    the temporary file in the same directory as finalPath.  If the code in
+    context succeeds, the file renamed to its actually name.  If an error
+    occurs, the file is not installed and is removed unless keep is specified.
+    """
+    def __init__(self, final_path, keep=False):
+        self.final_path = final_path
+        self.keep = keep
+        self.tmp_path = None
+
+    def __enter__(self):
+        self.tmp_path = atomic_tmp_file(self.final_path)
+        return self.tmp_path
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if exc_type is None:
+            atomic_install(self.tmp_path, self.final_path)
+        elif not self.keep:
+            try:
+                os.unlink(self.tmp_path)
+            except:
+                pass
+        return False
+
+def atomic_copy(src_path, dest_path):
+    """Copy a file using posix atomic creations semantics."""
+    with AtomicFileCreate(dest_path) as dest_path_tmp:
+        shutil.copyfile(src_path, dest_path_tmp)
```

Comment:
The word `atomic` does not appear in the documentation for `shutil.copyfile`, is there another place that this guarantee is made?

https://docs.python.org/3/library/shutil.html